Power Panic and LCD menu

3d-gussner opened the pull request:
Improved Power Panic
- Added Debug for printer states
- SD power panic
  - Recovers from multiple power panic and even resets (as long it is in recovering)
- Host power panic
- LCD menu
  - During recovering all menus are hidden except
    - Stop print (to be able to exit)
    - Statistics
    - Fail Stats
    - Support   
  - Reprint hidden
    - Host reprint only shown when host sends `M79` to ensure that the host is ready to reprint
    - SD Reprint only shown when previous print was finished successfully.
      - ToDo for next release: Don't show Reprint when SD card has been changed
  - Resume hidden
    - When resuming
    - With host prints until host sends `M79`
  - `Autoload filament` menu only shown with active FSensor and NO filament detected
  - `Load filament` menu only shown with active FSensor and NO filament detected OR FSensor disabled
  - `Unload filament` menu only shown with active FSensor and filament detected OR FSensor disabled
 
 
This Firmware doesn't recover correctly with Prusa Link 0.8.0-alpha4 at this moment.
I was able to recover host prints with a modified OctoPrint + two Plugins and additional code (which isn't in this PR as this code is a workaround of existing OctoPrint+Plugins)

Update 27-03-2024

Added `EEPROM_UVLO_Z_LIFTED` to prevent multiple Z lifts in case of a power panic while the print is paused.
A host print recovery it doesn't wait for hotend temp to be stable as it will go into pause state anyway and the hotend will be disabled while paused anyway. 

Tested with MK404

- [X] SD print + power panic
  - For some unknown reason it doesn't recover correctly SD print + power panic + reset while recovering, while a double power panic does recover correctly even with a reset.   
- [X] SD print + power panic and while lifting Z another power panic
- [X] SD print + power panic and while lifting Z another power panic + reset while recovering
- [X] SD print + pause + power panic
- [X] SD print + pause + power panic and while lifting Z another power panic
- [X] SD print + pause + power panic and while lifting Z another power panic + reset while recovering
- [X] Host print same as above for SD

Difference between SD and Host print power panic:
- SD recovers automatically when bed temp isn't lower than 5°C to target bed temperature
- Host recovers the print settings and goes to pause state.
  - Then it waits for the host sending "M79" before showing Resume to ensure that the host is "there" to receive any messages from the printer.  
   
The print can't be recovered if a reset happens at the last step of the recovery as at this moment the power panic recovery is "completed" and the printer is back in normal mode.

3d-gussner committed:
Add DEBUG_PRINTER_STATES

3d-gussner committed:
Add bool printer_recovering

3d-gussner committed:
Power panic recover lift by 20 Z_PAUSE_LIFT only once
by setting PENDING_RECOVERY_RETRY
Also the default saved_printing_type after a reboot is now PRINT_TYPE_NONE

3d-gussner committed:
Fix to correct EEPROM address

3d-gussner committed:
Cancle power recovery correctly using cancel_saved_printing();

3d-gussner committed:
Hide LCD menu "Reprint" during power panic recovery

3d-gussner committed:
Update debug_printer_states

3d-gussner committed:
Rename `sd_print_saved_in_ram` to `print_saved_in_ram` as it saves sd and host prints

3d-gussner committed:
Fix `print_saved_in_ram` to be true for both SD and host prints

3d-gussner committed:
Fix missing eeprom_update_byte for debuging

3d-gussner committed:
Change to saved_printing_type fixes few issues during power panic recovery

3d-gussner committed:
Fix another saved_printing_type issue

3d-gussner committed:
Add ubs_timer.start to PRUSA uvlo
Cleanup format

3d-gussner committed:
Reset EEPROM_UVLO after recovering
Undo PP Z lift

3d-gussner committed:
Fix DEBUG_UVLO_AUTOMATIC_RECOVER

3d-gussner committed:
LCD menu
- Hide menus during recovering
- Un-/Load filament with FSensor active only show Load OR Unload depending on FS sate
- Host
  - Reprint only shown when M79 timer is active
  - Resume only shown when M79 timer active
- Stop print always shown when printing, paused and recovering

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5228,6 +5228,10 @@ static void lcd_main_menu()
 #endif //RESUME_DEBUG
 
+#ifdef DEBUG_PRINTER_STATES
+    debug_printer_states();
```
I have a bad feeling about this. `lcd_main_menu()` is called very frequently and dumping this much stuff on the serial stream will likely do more harm than good. Does the printer actually keep up with this much serial prints?

gudnimg replied:
Another point is, what happens when the printer is not rendering the main menu? Then these debug prints are ignored.

3d-gussner replied:
You are correct, this menu will dump lot of information over serial and should only be used with the MK404 or with a subset of the messages.
I added this to be able to troubleshoot some LCD menu's.

gudnimg commented on Firmware/Marlin_main.cpp:
```diff
@@ -4117,62 +4173,52 @@ void process_commands()
-            if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_SD)
-            {
+        } else if (code_seen_P(PSTR("uvlo"))) { // PRUSA uvlo
+            if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_SD) {
```
Can `saved_printing_type` be used here?

`if (saved_printing_type == PowerPanic::PRINT_TYPE_SD)`

3d-gussner replied:
When you assume that `PRUSA uvlo` is only used by the "normal" Power Panic recovery functions then the `saved_printing_type` is already read before the `PRUSA uvlo`  is called.
But as this is a gcode that can be executed any time the `saved_printing_type` may not be correct.
I have no idea why this gcode even exists.

gudnimg commented on Firmware/Marlin_main.cpp:
```diff
@@ -4117,62 +4173,52 @@ void process_commands()
-            }
-            else if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_HOST)
-            {
+            } else if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_HOST) {
```
Can `saved_printing_type` be used here?

`if (saved_printing_type == PowerPanic::PRINT_TYPE_HOST)`

gudnimg commented on Firmware/Marlin_main.cpp:
```diff
@@ -11024,7 +11069,7 @@ void restore_print_from_ram_and_continue(float e_move)
 
     restore_print_file_state();
-
+    eeprom_update_byte((uint8_t*)EEPROM_UVLO, PowerPanic::NO_PENDING_RECOVERY);
 	lcd_setstatuspgm(MSG_WELCOME);
     saved_printing_type = PowerPanic::PRINT_TYPE_NONE;
 	saved_printing = false;
```
Optimisation idea: Can we use `cancel_saved_printing()` here?

So change:

```c
    eeprom_update_byte((uint8_t*)EEPROM_UVLO, PowerPanic::NO_PENDING_RECOVERY);
	lcd_setstatuspgm(MSG_WELCOME);
    saved_printing_type = PowerPanic::PRINT_TYPE_NONE;
	saved_printing = false;
```

to this:

```c
    cancel_saved_printing();
	lcd_setstatuspgm(MSG_WELCOME);
```

3d-gussner replied:
No as `cancle_saved_printing()` sets `saved_start_position[0] = SAVED_START_POSITION_UNSET;` and the function 
`void restore_print_from_ram_and_continue(float e_move)` only resets some states BUT continues to print.

gudnimg commented on Firmware/Marlin_main.cpp:
```diff
@@ -1581,32 +1636,33 @@ void setup()
-          #ifdef DEBUG_UVLO_AUTOMATIC_RECOVER 
+      //Restore printing type
+      saved_printing_type = eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE);
+#ifdef DEBUG_UVLO_AUTOMATIC_RECOVER
```
Personally I really don't like this `DEBUG_UVLO_AUTOMATIC_RECOVER` code. It makes the code very unreadable. We can probably move all the serial print messages into `recover_print()` function.

Just an observation/comment. Not a blocker.

3d-gussner replied:
It is 6 years old and as DEBUG code has been placed close to the operation you are trying to debug.
In VScode and an existing `Configuration_prusa.h` it is "grayed out" and so not that disturbing to read at least for me.

gudnimg commented on Firmware/power_panic.cpp:
```diff
@@ -315,11 +315,17 @@ void recover_print(uint8_t automatic) {
-    // and second also so one may remove the excess priming material.
+    // Undo PP Z Lift by setting current Z pos to + Z_PAUSE_LIFT
+    // After a reboot the printer doesn't know the Z height and we have to set its previous value
+    if(eeprom_read_byte((uint8_t*)EEPROM_UVLO) == PowerPanic::PENDING_RECOVERY_RETRY) {
+        current_position[Z_AXIS] += Z_PAUSE_LIFT;
+    }
+
```
I'm not sure about this change. `current_position` is likely not modified unless the g-code was executed.

`enquecommandf_P(PSTR("G1 Z%.3f F800"), current_position[Z_AXIS] + Z_PAUSE_LIFT);`

Doesn't the if statement below make sure this lift isn't executed again?:

`if(eeprom_read_byte((uint8_t*)EEPROM_UVLO) == PowerPanic::PENDING_RECOVERY)`

There shouldn't be another lift when the flag is set to `PowerPanic::PENDING_RECOVERY_RETRY`. Am I missing something?

3d-gussner replied:
Fixed with EEPROM_UVLO_Z_LIFTED

gudnimg reviewed (changes requested):
See my comments

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
     // only allow starting SD print if hardware errors (temperature or fan) are cleared
-    if(!get_temp_error()
-            && !printer_active()
+    if (!printer_recovering()
+        && !printer_active()
+        && !get_temp_error()
+        && !printer_active()
```
Redundant call to `!printer_active()` here

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
-        if ((nextSheet >= 0) && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
-            MENU_ITEM_FUNCTION_E(EEPROM_Sheets_base->s[sheet], eeprom_switch_to_next_sheet);
-        }
+        if(!printer_active()
+            && !farm_mode) {
```
Why is this split into two lines? Can easily fit one line and is more readable IMO

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
+            && !farm_mode) {
+            const int8_t sheet = eeprom_read_byte(&(EEPROM_Sheets_base->active_sheet));
+            const int8_t nextSheet = eeprom_next_initialized_sheet(sheet);
+            if ((nextSheet >= 0)
+                && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
```
Similar here. Why split this into two lines. I feel it is very unusual.
The comment can be either above the if statement or inside it.

3d-gussner committed:
Remove redundant check

3d-gussner committed:
Fix and improve PowerPanic
- No wait for heatup during host print recovery
- No extra Z lift when power panic happens during paused print
  - added an extra EEPROM value
  - fixed `EEPROM_LAST_ITEM`
  - changed EEPROM version to 3.14.0
- Added `reset_uvlo` function

3d-gussner committed:
Fix MK2.5 builds
- removed `reset_uvlo` as `#ifdef UVLO_SUPPORT` everywhere for MK2.5 is a huge change.

3d-gussner committed:
Revert split if statements
Kept few as these get too long

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
+                }
+            } else {
+                bMain=true;                                   // flag (i.e. 'fake parameter') for 'lcd_sdcard_menu()' function
+                MENU_ITEM_BACK_P(_i("No SD card")); ////MSG_NO_CARD c=18
```
This changes the UI, is this change needed?

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
     // only allow starting SD print if hardware errors (temperature or fan) are cleared
-    if(!get_temp_error()
-            && !printer_active()
+    if (!printer_recovering() && !printer_active() && !get_temp_error()
```
Potential optimisation: there are a few if statements inside this top-level if statements which check for `!printer_active()`. That shouldn't be needed anymore since this top-level if statement checks it.

3d-gussner replied:
Right, thanks

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
-        if ((nextSheet >= 0) && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
-            MENU_ITEM_FUNCTION_E(EEPROM_Sheets_base->s[sheet], eeprom_switch_to_next_sheet);
-        }
+        if(!printer_active() && !farm_mode) {
```
`!printer_active()` should always be true here since top-level if statement checks for it

3d-gussner replied:
Right, thanks

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
-                // The MMU 'Load filament' state machine will reject the command if any 
-                // filament sensor is reporting a detected filament
-                MENU_ITEM_SUBMENU_P(_T(MSG_PRELOAD_TO_MMU), mmu_preload_filament_menu);
+        if (!( printer_active() || (eFilamentAction != FilamentAction::None) || Stopped )) {
```
`!printer_active()` should always be true here since top-level if statement checks for it

3d-gussner replied:
Right, thanks

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -5249,45 +5253,43 @@ static void lcd_main_menu()
+        if (saved_printing_type == PowerPanic::PRINT_TYPE_SD) {
                 MENU_ITEM_SUBMENU_P(_T(MSG_RESUME_PRINT), lcd_resume_print);
-            } else {
+        } else if ((saved_printing_type == PowerPanic::PRINT_TYPE_HOST) && (M79_timer_get_status())) {
```
hmm, isn't this a regression for users which **don't** use the M79 timer? If a user doesn't use M79, this change means they can no longer resume the print from the LCD. But still they can pause the print from LCD.

See `lcd_pause_usb_print` which doesn't have this restriction.

3d-gussner replied:
It has been pointed out that the menu shouldn't be shown as long the host isn't connected.
Testing selected Resume too early while the RPi Zero was still booting up after a power panic. Nothing happened as the `//action:resume` never reached the host.
Using the `M79` ensures that the host is connected and ready to receive instructions like `//action:resume` and prevent users to select the menu too early.

gudnimg reviewed (changes requested):
See comments

3d-gussner committed:
Thanks to @gudnimg pointing out duplicate if statements that are covered by the top-level if statement.
